File: backend/app/Models/Task.php
<?php

namespace App\Models;

use App\Enums\TaskPriority;
use App\Enums\TaskStatus;
use Database\Factories\TaskFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

#[Fillable([
    'title', 'description', 'creator_id', 'organization_id', 'department_id',
    'task_category_id',
    'priority', 'status', 'start_date', 'due_date', 'completed_at',
    'progress', 'result',
])]
class Task extends Model
{
    /** @use HasFactory<TaskFactory> */
    use HasFactory;

    /**
     * Tasks this user may see in lists and counts: everything for a central
     * role; otherwise what they created or are assigned to, plus — with
     * `tasks.view` — their organization's tasks (only their department's
     * for a department-manager). `TaskPolicy::view()` mirrors this rule.
     */
    public function scopeVisibleTo(Builder $query, User $user): void
    {
        if ($user->hasCentralAccess()) {
            return;
        }

        $employee = $user->employee;

        $query->where(function ($involvedOrScoped) use ($user, $employee) {
            $involvedOrScoped->where('creator_id', $user->id)
                ->orWhereHas('assignees', fn ($assigneeQuery) => $assigneeQuery->where('employees.id', $employee?->id));

            if ($user->can('tasks.view')) {
                $involvedOrScoped->orWhere(function ($orgQuery) use ($user, $employee) {
                    $orgQuery->where('organization_id', $employee?->organization_id);

                    if ($user->hasRole('department-manager')) {
                        $orgQuery->where('department_id', $employee?->department_id);
                    }
                });
            }
        });
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(TaskCategory::class, 'task_category_id');
    }

    public function creator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'creator_id');
    }

    public function organization(): BelongsTo
    {
        return $this->belongsTo(Organization::class);
    }

    public function department(): BelongsTo
    {
        return $this->belongsTo(Department::class);
    }

    public function assignees(): BelongsToMany
    {
        return $this->belongsToMany(Employee::class, 'task_assignees')->withTimestamps();
    }

    public function comments(): HasMany
    {
        return $this->hasMany(TaskComment::class);
    }

    public function activities(): HasMany
    {
        return $this->hasMany(TaskActivity::class)->latest();
    }

    public function attachments(): HasMany
    {
        return $this->hasMany(TaskAttachment::class);
    }

    /**
     * Get the attributes that should be cast.
     *
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            // "date:Y-m-d", not plain "date" — see Employee::casts() for
            // why a bare "date" cast shifts across the day boundary once
            // JSON-serialized under a non-UTC APP_TIMEZONE.
            'start_date' => 'date:Y-m-d',
            'due_date' => 'date:Y-m-d',
            'completed_at' => 'datetime',
            'priority' => TaskPriority::class,
            'status' => TaskStatus::class,
            'progress' => 'integer',
        ];
    }
}

File: mobile/lib/features/tasks/domain/task.dart
import '../../../core/models/employee.dart';

class TaskComment {
  TaskComment({required this.id, this.userName, required this.body, required this.createdAt});

  final int id;
  final String? userName;
  final String body;
  final String createdAt;

  factory TaskComment.fromJson(Map<String, dynamic> json) => TaskComment(
        id: json['id'] as int,
        userName: json['user_name'] as String?,
        body: json['body'] as String? ?? '',
        createdAt: json['created_at'] as String? ?? '',
      );
}

class TaskAttachment {
  TaskAttachment({
    required this.id,
    required this.originalName,
    this.fileSize,
    required this.downloadUrl,
  });

  final int id;
  final String originalName;
  final int? fileSize;
  final String downloadUrl;

  factory TaskAttachment.fromJson(Map<String, dynamic> json) => TaskAttachment(
        id: json['id'] as int,
        originalName: json['original_name'] as String? ?? '',
        fileSize: json['file_size'] as int?,
        downloadUrl: json['download_url'] as String? ?? '',
      );
}

class TaskActivity {
  TaskActivity({required this.id, required this.action, this.description, this.causerName, required this.createdAt});

  final int id;
  final String action;
  final String? description;
  final String? causerName;
  final String createdAt;

  factory TaskActivity.fromJson(Map<String, dynamic> json) => TaskActivity(
        id: json['id'] as int,
        action: json['action'] as String? ?? '',
        description: json['description'] as String?,
        causerName: json['causer_name'] as String?,
        createdAt: json['created_at'] as String? ?? '',
      );
}

class Task {
  Task({
    required this.id,
    required this.title,
    this.description,
    this.creatorName,
    required this.assignees,
    required this.priority,
    required this.status,
    this.startDate,
    this.dueDate,
    this.completedAt,
    required this.progress,
    this.result,
    required this.comments,
    required this.attachments,
    required this.activities,
    required this.createdAt,
  });

  final int id;
  final String title;
  final String? description;
  final String? creatorName;
  final List<Employee> assignees;
  final String priority;
  final String status;
  final String? startDate;
  final String? dueDate;
  final String? completedAt;
  final int progress;
  final String? result;
  final List<TaskComment> comments;
  final List<TaskAttachment> attachments;
  final List<TaskActivity> activities;
  final String createdAt;

  bool get isActionable => status == 'new' || status == 'in_progress' || status == 'overdue';

  factory Task.fromJson(Map<String, dynamic> json) => Task(
        id: json['id'] as int,
        title: json['title'] as String? ?? '',
        description: json['description'] as String?,
        creatorName: json['creator_name'] as String?,
        assignees: (json['assignees'] as List<dynamic>? ?? [])
            .map((e) => Employee.fromJson(e as Map<String, dynamic>))
            .toList(),
        priority: json['priority'] as String? ?? 'normal',
        status: json['status'] as String? ?? 'new',
        startDate: json['start_date'] as String?,
        dueDate: json['due_date'] as String?,
        completedAt: json['completed_at'] as String?,
        progress: json['progress'] as int? ?? 0,
        result: json['result'] as String?,
        comments: (json['comments'] as List<dynamic>? ?? [])
            .map((e) => TaskComment.fromJson(e as Map<String, dynamic>))
            .toList(),
        attachments: (json['attachments'] as List<dynamic>? ?? [])
            .map((e) => TaskAttachment.fromJson(e as Map<String, dynamic>))
            .toList(),
        activities: (json['activities'] as List<dynamic>? ?? [])
            .map((e) => TaskActivity.fromJson(e as Map<String, dynamic>))
            .toList(),
        createdAt: json['created_at'] as String? ?? '',
      );
}
